flang/lib/Lower/OpenMP.cpp

```diff
@@ -1752,11 +1754,82 @@ bool ClauseProcessor::processMap(
                                        semanticsContext, stmtCtx, ompObject,
                                        clauseLocation, asFortran, bounds);
 
+          auto checkIfStructComponent =
+              [](const Fortran::parser::OmpObject &ompObject) {
+                bool isComponent = false;
+                std::visit(
+                    Fortran::common::visitors{
+                        [&](const Fortran::parser::Designator &designator) {
+                          if (auto *structComp = Fortran::parser::Unwrap<
+                                  Fortran::parser::StructureComponent>(
+                                  designator)) {
+                            if (std::holds_alternative<Fortran::parser::Name>(
+                                    structComp->base.u))
+                              isComponent = true;
+                          }
+                        },
+                        [&](const Fortran::parser::Name &name) {}},
+                    ompObject.u);
+
+                return isComponent;
+              };
+
+          // TODO: Currently, it appears there's missing symbol information
+          // and bounds information for allocatables and pointers inside
+          // of derived types. The latter needs some additional support
+          // added to the bounds generation whereas the former appears
+          // that it could be a problem when referring to pointer members
+          // via an OpenMP map clause, for the moment we do not handle
+          // these cases and must emit an error.
+          if (checkIfStructComponent(ompObject) &&
+              Fortran::semantics::IsAllocatableOrPointer(
+                  *getOmpObjectSymbol(ompObject)))
+            TODO(currentLocation,
+                 "pointer members of derived types are currently unmapped");
+
+          if (Fortran::semantics::IsAllocatableOrPointer(
+                  *getOmpObjectSymbol(ompObject))) {
+            // We mimic what will eventually be a structure containing a
+            // pointer mapping for allocatables/pointers/target e.g.:
+            //
+            // !$omp target map(from:in, in%map_ptr)
+            //
+            // ===>
+            //
+            // map_entry varptr(in) ....
+            // map_entry varptr(map_ptr) varptrptr(in) ...
```

Comment:
I like seeing the use of varPtrPtr since this is one aspect of OpenACC lowering which is not yet complete. :) However, I think there is a mismatch here - at least from how I thought about it when I added this operand in acc dialect. Consider how clang maps a pointer as array from a struct:
https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGOpenMPRuntime.cpp#L6989

Basically the varPtrPtr is literally meant to mean &varPtr. This means that this is the address of the "base_addr" field in the descriptor: aka (&(desc.base_addr)).
This is intentional - the whole descriptor does not need mapped if all we access is the data. Clang does it the same - the whole parent struct does not need mapped if only mapping a member.

Part of reason this wasn't solidified in lowering of OpenACC is because it is not yet clear how to really represent this abstraction. Descriptors don't have to be materialized into a struct type. And in that sense, in FIR, taking the address of a field in box type is not well defined.

Reply:
@jeanPerier @clementval I added you as reviewers to comment on this point. Is introducing a fir.box_addr_addr the right abstraction for solving this issue? How to ensure that such an operation is not abused or introduces any problems with alias analysis?

Reply:
Ah, then I misunderstood it then! Thank you for clarifying. Perhaps it'd be better creating a varParentPtr for OpenMP, if we continued down this route then, to seperate the two.

Yes, I thought we could probably ditch the descriptor at some point, but I thought of doing it as more of an optimisation case rather than the default, at least for the moment to get it working initially as it seemed simpler to get working initially whilst maintaining correctness at the trade off of it being slower for the time being. Simpler as the target region when lowered already takes into account the fact that the descriptor and pointer are lowered as a structure, so we simply treat it as such when mapping it. However, perhaps this isn't the best route to go down in the short-term or long-term? I've not yet looked at the complexity of implementing the other solution or the possible downsides, if any.

One case I can perhaps see that descriptors may be of use on the device is the following example:

```
      integer, target :: x(10)
      integer, pointer :: x_d(:)

      do i=1, 10
         x(i)=1
      enddo

      !$omp target map(from:x_d) map(to: x)
       x_d => x;
      !$omp end target
```

Where the descriptor for the pointer may need updated so that the host can appropriately utilise it, but this is just a thought and probably not a very well thought out one at that.

Reply:
I admit I don't see why create a varParentPtr instead of varPtrPtr. You can deduce the actual parent from the varPtrPtr since an access of a member (when it gets materialized). Also, if you don't have the access to the field directly how do you know how to do the attach action? Interestingly in this case, &descriptor == &(descriptor.base_addr), so that is probably why it works just fine for LLVM - but again I feel like varParentPtr is not the abstraction you really want.

Reply:
You're likely correct, what would you suggest is the correct abstraction in this case? 

I'm unsure what you mean by attach in this case unfortunately, so I apologies. But I likely don't do it correctly, so I would very much love to hear the correct direction you think it should go so I can try to adjust it to that! 

I can open a patch on the lowering that I have for it in the next few days hopefully if that'd help clarify anything/aid discussion.

Reply:
Sorry about using "attach" terminology - after looking through OpenMP spec, it does not seem to be the common use of the semantics I meant. https://www.openmp.org/spec-html/5.2/openmpsu6.html#x13-120001.2.6 look for "attached pointer".
Anyway, attach in my use of it, means, that if you have something like:
struct S {
  double * p;
} s;
map(tofrom:s,s.p)
Then in device memory, s is slightly different. Its field `p` points to the device memory. The concept of changing device `s`'s p field is called `attach` in OpenACC terms.

Reply:
I think fir.box_addr_addr would likely be OK if it helps here. In terms of side-effects/alias analysis it would be like a fir.coordinate or a GEP.

I would however not introduce similar operations for other descriptor fields like extents/bounds. It is safe to require fir.box codegen to include an address field somewhere, but it is not safe to assume that it will hold an extent field for instance (some descriptor format store the upper bound and recompute the extent when needed or vice-versa).

Reply:
> Sorry about using "attach" terminology - after looking through OpenMP spec, it does not seem to be the common use of the semantics I meant. https://www.openmp.org/spec-html/5.2/openmpsu6.html#x13-120001.2.6 look for "attached pointer". Anyway, attach in my use of it, means, that if you have something like: struct S { double * p; } s; map(tofrom:s,s.p) Then in device memory, s is slightly different. Its field `p` points to the device memory. The concept of changing device `s`'s p field is called `attach` in OpenACC terms.

No need to apologies, even if you used the OpenMP terminology I wouldn't have understood in this case, so thank you for looking it up for me and teaching me in this case. 

To perhaps give a bit more information to hopefully aid discussion, I believe it works as I do some cheating to recreate the GEP for the final segment in this example: https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CGOpenMPRuntime.cpp#L6989 that likely only works in this specific case and I'd love to be able to fix it as it's on my TODO list! On the off chance that you're interested in looking at the lowering to LLVM-IR that I have in place currently, here's the little blob of code that I've currently managed to put together to generate the relevant runtime components to do the initial offload of the allocatables/pointers as a special case of a structure (sorry it's downstream and a tad messy, you can ignore effectively everything outside of this function):  https://github.com/ROCm-Developer-Tools/llvm-project/pull/225/files#diff-2cbb5651f4570d81d55ac4198deda0f6f7341b2503479752ef2295da3774c586R1983

The varPtrPtr being the parent in this perhaps terrible use-case is primarily to indicate that something is part of another object and to then sort it into groupings of parent + children so they can be mapped as such. Using the varPtrPtr for this is likely not the best way to do things and is as you've stated a misuse of the field (hence the suggestion of a parent field, but that in itself is perhaps misguided thinking). 

This is initial/WIP work as someone new to a lot of OpenMP/OpenACC and Fortran concepts, so I am more than happy to adjust/change things based on suggestions and look into other directions if there's better avenues! e.g. perhaps as I think you've said mapping the descriptor is not the way to go long term (although, for initial functionality it could be nice to treat them as such, with descriptor removal being a later step and more optimal approach).

Thank you for your input on this patch as well as @jeanPerier it's very appreciated.

Reply:
> I think fir.box_addr_addr would likely be OK if it helps here. In terms of side-effects/alias analysis it would be like a fir.coordinate or a GEP.
> 
> I would however not introduce similar operations for other descriptor fields like extents/bounds. It is safe to require fir.box codegen to include an address field somewhere, but it is not safe to assume that it will hold an extent field for instance (some descriptor format store the upper bound and recompute the extent when needed or vice-versa).

Adding a whole new operation might be a little overkill if it's only useful for this specific situation, I imagine, so unless it's absolutely necessary I'd not want to impose something like that on the FIR Dialect, but if it has other use cases/usages then perhaps it'd be good to add!